File: DailyPractice/20240309/LC_719.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        int lo=0,hi=1e6,ans=hi;

        sort(nums.begin(),nums.end());
        int n=nums.size();
        
        while(hi>=lo){
            int mid=(hi+lo)/2;
            int cnt=0;

            //least number where cnt of pairs whose distance <= to the number

            for(int i=0;i<n;i++){
                int c=upper_bound(nums.begin(),nums.end(),nums[i]+mid)-nums.begin();
                c--;
                cnt=cnt+max(0,c-i);
            }

            if(cnt>=k){ 
                ans=min(ans,mid);
                hi=mid-1;
            }else{
                lo=mid+1;
            }
        }

        return ans;
    }
};

File: DailyPractice/20240814/LC_719.java


public class LC_719 {

    class Solution {
        public int smallestDistancePair(int[] nums, int k) {
            
            int n=nums.length;
            Arrays.sort(nums);
            int lo=0,hi=1000000,ans=hi;
    
            while(hi>=lo){
                int mid=(hi+lo)/2;
                int j=0,cnt=0;
    
                for(int i=0;i<n;i++){
                    while(j<i && nums[i]-nums[j]>mid){
                        j++;
                    }
                    cnt=cnt+(i-j);
                }
    
    
                if(cnt>=k){
                    ans=Math.min(ans,mid);
                    hi=mid-1;
                }else{
                    lo=mid+1;
                }
    
            }
    
            return ans;
    
        }
    }
    
}
